src/server/dflycmd.cc

```diff
@@ -431,49 +431,53 @@ void DflyCmd::TakeOver(CmdArgList args, ConnectionContext* cntx) {
   }
   VLOG(1) << "AwaitCurrentDispatches done";
 
-  // We have this guard to disable expirations: We don't want any writes to the journal after
-  // we send the `PING`, and expirations could ruin that.
-  // TODO: Decouple disabling expirations from TransactionGuard because we don't
-  // really need TransactionGuard here.
-  TransactionGuard tg{cntx->transaction, /*disable_expirations=*/true};
-
-  if (*status == OpStatus::OK) {
-    auto cb = [&cntx = replica_ptr->cntx, replica_ptr = replica_ptr, timeout_dur, start,
-               &status](EngineShard* shard) {
-      FlowInfo* flow = &replica_ptr->flows[shard->shard_id()];
-
-      shard->journal()->RecordEntry(0, journal::Op::PING, 0, 0, {}, true);
-      while (flow->last_acked_lsn < shard->journal()->GetLsn()) {
-        if (absl::Now() - start > timeout_dur) {
-          LOG(WARNING) << "Couldn't synchronize with replica for takeover in time: "
-                       << replica_ptr->address << ":" << replica_ptr->listening_port
-                       << ", last acked: " << flow->last_acked_lsn << ", expecting "
-                       << shard->journal()->GetLsn();
-          status = OpStatus::TIMED_OUT;
-          return;
-        }
-        if (cntx.IsCancelled()) {
-          status = OpStatus::CANCELLED;
-          return;
+  {
+    // We have this guard to disable expirations: We don't want any writes to the journal after
+    // we send the `PING`, and expirations could ruin that.
+    // TODO: Decouple disabling expirations from TransactionGuard because we don't
+    // really need TransactionGuard here.
+    TransactionGuard tg{cntx->transaction, /*disable_expirations=*/true};
```

Comment:
I am not sure adding this inside scope is correct. we have this to disable explorations while takeover. I dont fully remember all the details here but it might be that we will send again expirations to replica after we exit this scope? if so this is wrong

Reply:
the issue is that we deadlock when using the same transaction guard in do save... I don't remember about expiration, but we generally shouldn't be running them in a non-active state 🙂 

The other more solution is keeping the code as is and just using an artificial multi transaction for this....

Reply:
why we did not have this deadlock before? as we called save from ShutdownCmd.
Any way we can solve this by simply extracting the implementation of disable expiration out of TransactionGuard, as you can see in the comment we dont really need a TransactionGuard but need to disable expiration

Reply:
> why we did not have this deadlock before?

because we don't shut down the service there, we only request it (How could with with dispatch in progress 🙂). It happens from dfly_main